AsyncDisplayKit/Layout/ASStackLayoutSpec.mm

```diff
@@ -120,43 +122,62 @@ - (void)setBaselineRelativeArrangement:(BOOL)baselineRelativeArrangement
 
 - (ASLayout *)calculateLayoutThatFits:(ASSizeRange)constrainedSize
 {
-  std::vector<id<ASLayoutElement>> stackChildren;
-  for (id<ASLayoutElement> child in self.children) {
-    stackChildren.push_back(child);
-  }
-  
-  if (stackChildren.empty()) {
+  NSArray *children = self.children;
+  if (children.count == 0) {
     return [ASLayout layoutWithLayoutElement:self size:constrainedSize.min];
   }
+ 
+  // Accessing the style and size property is pretty costly we create layout spec children we use to figure
+  // out the layout for each child
+  const auto stackChildren = AS::map(children, [&](const id<ASLayoutElement> child) -> ASStackLayoutSpecChild {
+    ASLayoutElementStyle *style = child.style;
+    return {child, style, style.size};
+  });
   
-  ASStackLayoutSpecStyle style = {.direction = _direction, .spacing = _spacing, .justifyContent = _justifyContent, .alignItems = _alignItems, .baselineRelativeArrangement = _baselineRelativeArrangement};
-  BOOL needsBaselinePass = _baselineRelativeArrangement || _alignItems == ASStackLayoutAlignItemsBaselineFirst || _alignItems == ASStackLayoutAlignItemsBaselineLast;
+  // TOOD: Does this needs locking?
+  const ASStackLayoutSpecStyle style = {.direction = _direction, .spacing = _spacing, .justifyContent = _justifyContent, .alignItems = _alignItems, .baselineRelativeArrangement = _baselineRelativeArrangement};
   
+  // First pass is to get the children into a positioned state
   const auto unpositionedLayout = ASStackUnpositionedLayout::compute(stackChildren, style, constrainedSize);
   const auto positionedLayout = ASStackPositionedLayout::compute(unpositionedLayout, style, constrainedSize);
   
+  // Figure out if a baseline pass is really needed
+  const BOOL directionIsVertical = (style.direction == ASStackLayoutDirectionVertical);
+  const BOOL needsBaselineAlignment = ASStackBaselinePositionedLayout::needsBaselineAlignment(style);
+  const BOOL needsBaselinePositioning = (directionIsVertical == NO || needsBaselineAlignment == YES);
+  
+  NSMutableArray *sublayouts = [NSMutableArray array];
   CGSize finalSize = CGSizeZero;
-  NSArray *sublayouts = nil;
+  if (needsBaselinePositioning) {
+    // Regardless of whether or not this stack aligns to baseline, we need to let ASStackBaselinePositionedLayout::compute find the max ascender
```

Comment:
This comment is a little misleading now. Let's change it to something like:

All horizontal stacks, regardless of whether or not they are baseline aligned, should go through a baseline computation. They could be used in another horizontal stack that is baseline aligned and will need to have computed the proper ascender/descender. 

Vertical stacks do not need to go through this computation since we can easily compute ascender/descender by looking at their first/last child's ascender/descender.